### PR TITLE
[jdatabase] -  importer exporter

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
this is the continuation of the work  made by @Hackwar  #10991

- added the driver pdomysql,postgresql,sqlsrv
- added a cli script for testing purpose and as an example of how to use

### Testing Instruction
you can run the cli script to export tables

- `php exporter.php --all --folder  <folder_path> `    
   export all tables and data to the folder        

- `php exporter.php --all --mode zip`                
   export all tables and data in zip file

- `php exporter.php --table <table_name>`
   export table name

- `php exporter.php --table <table_name> --mode zip`
   export table name as a .zip file

you can run the cli script to import tables

- `php importer.php --all --folder <folder_path>`             
import all from folder
- `php importer.php --table <table_name>`
import table

### CODE REVIEW

File: cli/exporter.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		$pathPart = JPATH_ROOT . '/cli/dbdump/';
+		$tables   = JFactory::getDbo()->getTableList();
+		$prefix   = JFactory::getDbo()->getPrefix();
+		$zipfile  = 'jdata_exported_' . JFactory::getDate()->format('Y-m-d') . '.zip';
+
+		$exp     = JFactory::getDbo()->getExporter()->withStructure();
+		$iall    = $this->input->getString('all', false);
+		$ihelp   = $this->input->getString('help', false);
+		$itable  = $this->input->getString('table', false);
+		$imode   = $this->input->getString('mode', false);
```
Comment (yvesh) [inline cli/exporter.php]: you want to have a string, so null as default ;) This is also false in your if

File: cli/exporter.php [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		}
+
+		$zip = JArchive::getAdapter('zip');
+
+		foreach ($tables as $table)
+		{
+			if (strpos(substr($table, 0, strlen($prefix)), $prefix) !== false)
+			{
+				$task_i_time = microtime(true);
+				$filename = $pathPart . $table . '.xml';
```
Comment (yvesh) [inline cli/exporter.php]: code style

File: cli/exporter.php (hunk 1 above)
Comment (alikon) [reply cli/exporter.php]: fixed null as defualt

File: cli/importer.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		$this->out('============================');
+		$total_time = microtime(true);
+
+		// Import the dependencies
+		jimport('joomla.filesystem.file');
+		jimport('joomla.filesystem.folder');
+		$pathPart = JPATH_ROOT . '/cli/dbdump/';
+		$iall    = $this->input->getString('all', false);
+		$ihelp   = $this->input->getString('help', false);
+		$itable  = $this->input->getString('table', false);
```
Comment (yvesh) [inline cli/importer.php]: same here - null

File: cli/exporter.php (hunk 2 above)
Comment (alikon) [reply cli/exporter.php]: fixed

File: cli/exporter.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	 */
+	public function doExecute()
+	{
+		$this->out(JText::_('DbExporterCli'));
+		$this->out('============================');
+		$total_time = microtime(true);
+
+		// Import the dependencies
+		jimport('joomla.filesystem.file');
+		$pathPart = JPATH_ROOT . '/cli/dbdump/';
```
Comment (yvesh) [inline cli/exporter.php]: hmm the thing here is, you are putting sql dumps in a public accessible web directory... At least warn the user that he needs to take it out!

Comment (yvesh) [reply cli/exporter.php]: Or even better, allow the user to set the path, so he can take it outside his webroot

File: cli/importer.php (hunk 3 above)
Comment (alikon) [reply cli/importer.php]: fixed

File: libraries/joomla/database/exporter/mysqli.php [hunk 5]
```diff
@@ -85,6 +94,75 @@ protected function buildXmlStructure()
...
+			$this->db->setQuery($query);
+
+			$rows = $this->db->loadObjectList();
+
+			if (!count($rows))
+			{
+				continue;
+			}
+
+			$buffer[] = '  <table_data name="' . $table . '">';
```
Comment (yvesh) [inline libraries/joomla/database/exporter/mysqli.php]: Hmm did you check other options to serialize php objects to XML, like XML_Serializer? Not a fan of generate hard-coded XML through php..

File: libraries/joomla/database/exporter/pdomysql.php [hunk 6]
```diff
@@ -87,6 +96,75 @@ protected function buildXmlStructure()
...
+			$this->db->setQuery($query);
+
+			$rows = $this->db->loadObjectList();
+
+			if (!count($rows))
+			{
+				continue;
+			}
+
+			$buffer[] = '  <table_data name="' . $table . '">';
```
Comment (yvesh) [inline libraries/joomla/database/exporter/pdomysql.php]: Too much duplicate code ;)

File: libraries/joomla/database/exporter/postgresql.php [hunk 7]
```diff
@@ -101,6 +117,56 @@ protected function buildXmlStructure()
...
+				->from($query->qn($table));
+			$this->db->setQuery($query);
+			$rows = $this->db->loadObjectList();
+
+			if (!count($rows))
+			{
+				continue;
+			}
+
+			$buffer[] = '  <table_data name="' . $table . '">';
```
Comment (yvesh) [inline libraries/joomla/database/exporter/postgresql.php]: Make at least a helper :P

File: libraries/joomla/database/exporter/sqlsrv.php [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+			$rows = $this->db->loadObjectList();
+
+			if (!count($rows))
+			{
+				continue;
+			}
+
+			$buffer[] = '  <table_data name="' . $table . '">';
+
+			foreach ($rows as $row)
```
Comment (yvesh) [inline libraries/joomla/database/exporter/sqlsrv.php]: That's still to much duplicate code :D

File: libraries/joomla/database/importer/mysqli.php [hunk 9]
```diff
@@ -416,7 +424,8 @@ protected function getKeySql($columns)
 		$kNonUnique = (string) $columns[0]['Non_unique'];
 		$kName = (string) $columns[0]['Key_name'];
 		$kColumn = (string) $columns[0]['Column_name'];
-
+		$kLength = (string) $columns[0]['Sub_part'];
+		$kLength = $kLength == '' ? '' : '(' . $kLength . ')';
```
Comment (yvesh) [inline libraries/joomla/database/importer/mysqli.php]: use ?:

File: libraries/joomla/database/importer/mysqli.php [hunk 10]
```diff
@@ -433,13 +442,15 @@ protected function getKeySql($columns)
...
-			$kColumns[] = $this->db->quoteName($kColumn);
+			$kColumns[] = $this->db->quoteName($kColumn) . $kLength;
 		}
 		else
 		{
 			foreach ($columns as $column)
 			{
-				$kColumns[] = (string) $column['Column_name'];
+				$kLength = (string) $column['Sub_part'];
+				$kLength = $kLength == '' ? '' : '(' . $kLength . ')';
```
Comment (yvesh) [inline libraries/joomla/database/importer/mysqli.php]: ?:

File: libraries/joomla/database/importer/pdomysql.php [hunk 11]
```diff
@@ -379,6 +375,8 @@ protected function getKeySql($columns)
 		$kNonUnique = (string) $columns[0]['Non_unique'];
 		$kName      = (string) $columns[0]['Key_name'];
 		$kColumn    = (string) $columns[0]['Column_name'];
+		$kLength = (string) $columns[0]['Sub_part'];
+		$kLength = $kLength == '' ? '' : '(' . $kLength . ')';
```
Comment (yvesh) [inline libraries/joomla/database/importer/pdomysql.php]: ?:

File: libraries/joomla/database/importer/pdomysql.php [hunk 12]
```diff
@@ -395,13 +393,15 @@ protected function getKeySql($columns)
...
-			$kColumns[] = $this->db->quoteName($kColumn);
+			$kColumns[] = $this->db->quoteName($kColumn) . $kLength;
 		}
 		else
 		{
 			foreach ($columns as $column)
 			{
-				$kColumns[] = (string) $column['Column_name'];
+				$kLength = (string) $column['Sub_part'];
+				$kLength = $kLength == '' ? '' : '(' . $kLength . ')';
```
Comment (yvesh) [inline libraries/joomla/database/importer/pdomysql.php]: ?:

File: libraries/joomla/database/exporter.php [hunk 13]
```diff
@@ -244,4 +244,73 @@ public function withData($setting = false)
 
 		return $this;
 	}
+
+	/**
+	 * Builds the XML data to export.
+	 *
+	 * @return  array  An array of XML lines (strings).
+	 *
+	 * @since   3.6
```
Comment (zero-24) [inline libraries/joomla/database/exporter.php]: `__DEPLOY_VERSION__` please ;)

File: cli/exporter.php (hunk 4 above)
Comment (alikon) [reply cli/exporter.php]: added a `--folder`   paramametr

File: libraries/joomla/database/exporter/sqlsrv.php (hunk 8 above)
Comment (alikon) [reply libraries/joomla/database/exporter/sqlsrv.php]: duplicate code re-moved

File: libraries/joomla/database/importer/mysqli.php (hunk 9 above)
Comment (alikon) [reply libraries/joomla/database/importer/mysqli.php]: code rewritten
thx

File: libraries/joomla/database/exporter.php (hunk 13 above)
Comment (alikon) [reply libraries/joomla/database/exporter.php]: changed

File: libraries/joomla/database/exporter/sqlsrv.php [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+ * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE
+ */
+
+defined('JPATH_PLATFORM') or die;
+
+/**
+ * Sqlsrv export driver.
+ *
+ * @since  11.1
```
Comment (zero-24) [inline libraries/joomla/database/exporter/sqlsrv.php]: please in this file too ;)

File: libraries/joomla/database/importer.php [hunk 15]
```diff
@@ -153,6 +153,51 @@ protected function getRealTableName($table)
 	}
 
 	/**
+	 * Import the data from the source into the existing tables.
+	 *
+	 * @return  void
+	 *
+	 * @note    Currently only supports XML format.
+	 * @since   3.6
```
Comment (zero-24) [inline libraries/joomla/database/importer.php]: and here

File: libraries/joomla/database/importer/pdomysql.php [hunk 16]
```diff
@@ -434,4 +424,47 @@ public function check()
...
 	}
+
+	/**
+	 * Get the SQL syntax to add a table.
+	 *
+	 * @param   SimpleXMLElement  $table  The table information.
+	 *
+	 * @return  string
+	 *
+	 * @since   11.1
```
Comment (zero-24) [inline libraries/joomla/database/importer/pdomysql.php]: and here

File: libraries/joomla/database/importer/postgresql.php [hunk 17]
```diff
@@ -42,6 +42,86 @@ public function check()
 	}
 
 	/**
+	 * Get the SQL syntax to add a table.
+	 *
+	 * @param   SimpleXMLElement  $table  The table information.
+	 *
+	 * @return  string
+	 *
+	 * @since   11.1
```
Comment (zero-24) [inline libraries/joomla/database/importer/postgresql.php]: and here

File: libraries/joomla/database/importer/postgresql.php [hunk 18]
```diff
@@ -42,6 +42,86 @@ public function check()
...
+	}
+
+	/**
+	 * Get the details list of keys for a table.
+	 *
+	 * @param   array  $keys  An array of objects that comprise the keys for the table.
+	 *
+	 * @return  array  The lookup array. array({key name} => array(object, ...))
+	 *
+	 * @since   11.1
```
Comment (zero-24) [inline libraries/joomla/database/importer/postgresql.php]: and here

**Review by zero-24 — DISMISSED** [review]
`__DEPLOY_VERSION__` ;)

**Comment by zero-24** [comment]
@alikon i have found a few more places ;)

File: libraries/joomla/database/exporter/sqlsrv.php (hunk 14 above)
Comment (alikon) [reply libraries/joomla/database/exporter/sqlsrv.php]: done

File: libraries/joomla/database/importer.php (hunk 15 above)
Comment (alikon) [reply libraries/joomla/database/importer.php]: done

File: libraries/joomla/database/importer/pdomysql.php (hunk 16 above)
Comment (alikon) [reply libraries/joomla/database/importer/pdomysql.php]: done

File: libraries/joomla/database/importer/postgresql.php (hunk 17 above)
Comment (alikon) [reply libraries/joomla/database/importer/postgresql.php]: done

File: libraries/joomla/database/importer/postgresql.php (hunk 18 above)
Comment (alikon) [reply libraries/joomla/database/importer/postgresql.php]: done

**Comment by alikon** [comment]
@zero-24 should be fixed now

File: libraries/joomla/database/importer/sqlsrv.php [hunk 19]
```diff
@@ -0,0 +1,435 @@
...
+					for ($i = 0; $i < $newCount; $i++)
+					{
+						$same = (((string) $newLookup[$name][$i]['Non_unique'] == $oldLookup[$name][$i]->Non_unique)
+							&& ((string) $newLookup[$name][$i]['Column_name'] == $oldLookup[$name][$i]->Column_name)
+							&& ((string) $newLookup[$name][$i]['Seq_in_index'] == $oldLookup[$name][$i]->Seq_in_index)
+							&& ((string) $newLookup[$name][$i]['Collation'] == $oldLookup[$name][$i]->Collation)
+							&& ((string) $newLookup[$name][$i]['Index_type'] == $oldLookup[$name][$i]->Index_type));
+
+						/*
+						Debug.
```
Comment (zero-24) [inline libraries/joomla/database/importer/sqlsrv.php]: debug code? ;)

**Comment by zero-24** [comment]
@alikon i have fixed some missing deploy version things ;)

File: libraries/joomla/database/importer/sqlsrv.php (hunk 19 above)
Comment (alikon) [reply libraries/joomla/database/importer/sqlsrv.php]: removed

**Comment by infograf768** [comment]
Should https://github.com/joomla/joomla-cms/pull/10991 be closed ?

**Comment by alikon** [comment]
can be closed cause this pr include all from #10991

**Comment by BombatiK** [comment]
I have tested this item :red_circle: unsuccessfully on 696715d3c861e8d80055d906b513989b6a6ccedd<br /><br />i have tested it giving my own folder_path, but it always create a folder dbdump where it sets the result. 
Is it the right way?
@icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14272">issues.joomla.org/tracker/joomla-cms/14272</a>.</sub>

**Comment by jonasgonka** [comment]
I have tested this item :white_check_mark: successfully on 696715d3c861e8d80055d906b513989b6a6ccedd<br /><br />Tested successfully on Windows.

Command `php exporter.php --all  --folder C:\Users\user\Desktop\tables\` created the folder `tables` as expected.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14272">issues.joomla.org/tracker/joomla-cms/14272</a>.</sub>

**Comment by alikon** [comment]
@BombatiK can you show me the command that you have  used ?

**Comment by BombatiK** [comment]
@alikon 
php exporter.php --all --folder C:\xampp\htdocs\joomla2\cli\db

@icampus

**Comment by alikon** [comment]
sorry  but don't understand  is the same command issued by @jonasgonka

**Comment by jonasgonka** [comment]
I think the problem is that he is missing the last backslash.
Maybe it would be better to switch to using actual directories in the script instead of just the strings.

**Comment by twister65** [comment]
Exporter works fine (without error) when I include a slash (backslash with Window) at the end of the folder.
Importer gives no error.
But after all, when I try to write in the database, creating an article for example, I have this error : 

> Error
> Save failed with the following error: ERROR: null value in column "id" violates not-null constraint

I did my test with the postgresql driver.

**Comment by twister65** [comment]
Is it possible to use the importer directly with a zip file previously created by the exporter ?

**Comment by twister65** [comment]
Importer failed with the PostgreSQL PDO driver : 

> DbImporterCli
============================
Importing gvcms_action_log_config from gvcms_action_log_config.xml
Reading data from gvcms_action_log_config.xml
Drop gvcms_action_log_config
Error on mergeStructuregvcms_action_log_config.xml JDatabaseExceptionExecuting: 00000, ,  in /var/www/html/avenirdurable/libraries/joomla/database/driver/pdo.php:475
Stack trace:
#0 /var/www/html/avenirdurable/libraries/joomla/database/importer.php(251): JDatabaseDriverPdo->execute()
#1 /var/www/html/avenirdurable/cli/importer.php(139): JDatabaseImporter->mergeStructure()
#2 /var/www/html/avenirdurable/libraries/src/Application/CliApplication.php(143): DbImporterCli->doExecute()
#3 /var/www/html/avenirdurable/cli/importer.php(170): Joomla\CMS\Application\CliApplication->execute()
#4 {main}

**Comment by twister65** [comment]
@jonasgonka , could you specify which database driver you used ?

**Comment by twister65** [comment]
https://github.com/joomla/joomla-cms/pull/14272#issuecomment-445517875

**Structure parameters are missing.**

gvcms_content_id_seq is incorrect in the postgresql database.
We should have yes (instead of no) for 

> Will increment last value before returning next value (is_called)?

Last value is 1 instead of 30 (in my test case).

In addition, the gvcms_content table is incomplete in the postgresql database. We have no default value. For id, this should be:

> nextval('public.gvcms_content_id_seq'::regclass)

**Comment by twister65** [comment]
Please, replace (`$field->default`):
https://github.com/joomla/joomla-cms/blob/786aa3d681dba5a6392f10766f8071f43da8a112/libraries/joomla/database/exporter/postgresql.php#L100
by (`$field->Default`):
`							(isset($field->Default) ? ' Default="' . $field->Default . '"' : '') . ' Comments="' . $field->comments . '"' .`

It fixes (https://github.com/joomla/joomla-cms/pull/14272#issuecomment-445549812):

> We have no default value.

**Comment by twister65** [comment]
We could add `Is_called="YES` at the sequence:
https://github.com/joomla/joomla-cms/blob/786aa3d681dba5a6392f10766f8071f43da8a112/libraries/joomla/database/exporter/postgresql.php#L92
`' Cycle_option="' . $sequence->cycle_option . '"' . ' Is_called="YES"' .`

**Comment by twister65** [comment]
We have to clean default value, adding:
`		
                $cDefault = strstr($field['Default'], '::', true);
`
`
		if ($cDefault != false)
		{
			$field['Default'] = $cDefault;
		}
`
before:
https://github.com/joomla/joomla-cms/blob/786aa3d681dba5a6392f10766f8071f43da8a112/libraries/joomla/database/importer/postgresql.php#L487-L489

**Comment by twister65** [comment]
Modify : 
https://github.com/joomla/joomla-cms/blob/786aa3d681dba5a6392f10766f8071f43da8a112/libraries/joomla/database/importer/postgresql.php#L365
		return 'CREATE SEQUENCE IF NOT EXISTS ' . (string) $field['Name'] .

**Comment by alikon** [comment]
@twister65  sorry for not be so responsive in these days, 
yes the postgresql import/export need a little bit more care, hope i should be more free in the next weeks

**Comment by twister65** [comment]
PostgreSQL.
The last value in the sequences is not defined (1). When a new row is added to a table, the id is incorrect (1).

**Comment by twister65** [comment]
Thanks @alikon , could you modify the cli script to add a driver parameter ? 
For example, migrate a PostgreSQL database (export) to a MySQL database (import).
I appreciate this PR, especially since Akeeba Backup does not support PostgreSQL.
In addition, I plan to migrate my PostgreSQL database to MySQL.
I hope it will work well someday. I expect a lot from this PR.
Regards.

**Comment by brianteeman** [comment]
> I appreciate this PR, especially since Akeeba Backup does not support PostgreSQL.

It does

**Comment by twister65** [comment]
Akeeba Database backup engine is : 

- Native MySQL backup engine
or
- Reverse engineering database dump engine.

There is no Native PostgreSQL engine. I never succeeded to make reverse engineering work.
Instead I use pg_dump to make a backup.

**Comment by alikon** [comment]
closed in favour of  #23320

**Pull request closed without merging**